test/OpenTelemetry.Instrumentation.GrpcClient.Tests/GrpcClientTests.cs

```diff
@@ -0,0 +1,107 @@
+﻿// <copyright file="GrpcClientTests.cs" company="OpenTelemetry Authors">
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// </copyright>
+using System;
+using System.Diagnostics;
+using System.Linq;
+using System.Net;
+using System.Threading.Tasks;
+using Greet;
+using Grpc.Core;
+using Grpc.Net.Client;
+using Moq;
+using OpenTelemetry.Instrumentation.GrpcClient.Internal.Tests;
+using OpenTelemetry.Trace;
+using OpenTelemetry.Trace.Configuration;
+using OpenTelemetry.Trace.Export;
+using Xunit;
+
+namespace OpenTelemetry.Instrumentation.GrpcClient.Tests
+{
+    public class GrpcClientTests
+    {
+        [Theory]
+        [InlineData("https://localhost")]
+        [InlineData("https://127.0.0.1")]
+        [InlineData("https://[::1]")]
+        public async Task GrpcClientCallsAreCollectedSuccessfully(string baseAddress)
+        {
+            var uri = new Uri(baseAddress);
+            var uriHostNameType = Uri.CheckHostName(uri.Host);
+
+            var spanProcessor = new Mock<ActivityProcessor>();
+
+            var parent = new Activity("parent")
+                .SetIdFormat(ActivityIdFormat.W3C)
+                .Start();
+
+            using (OpenTelemetrySdk.EnableOpenTelemetry(
+                (builder) => builder
+                    .AddGrpcClientDependencyInstrumentation()
```

Comment:
Can you add a test with AddDependencyInstrumentation(), and see if we can trigger the activity from both grpcclient and its internal httpclient? We can address that scenario in future, but having a test here would show the behaviour.

Reply:
I like this idea. Currently, the test scaffolding defines its own `HttpMessageHandler` which means the "real" `HttpClient.SendAsync` does not get invoked. Though, I think there's probably a way I can wire things up to utilize the `TestHttpServer`. I'll try this.

Reply:
I've added a test that uses `AddDependencyInstrumentation()` and I also greatly simplified the test scaffolding.

However, I do have a flakey test now that I am investigating... seems the grpc span sometimes does not get end in time.

Reply:
I figured out the source of the flakiness.

Starting from the client call in the test: `client.SayHello(new HelloRequest());`
* `HttpClientCallInvoker.BlockingUnaryCall` is invoked [here](https://github.com/grpc/grpc-dotnet/blob/v2.25.x/src/Grpc.Net.Client/Internal/HttpClientCallInvoker.cs#L113).
* `HttpClientCallInvoker.BlockingUnaryCall` blocks on the `ResponseAsync` task which is completed in `GrpcCall.RunCall` [here](https://github.com/grpc/grpc-dotnet/blob/v2.25.x/src/Grpc.Net.Client/Internal/GrpcCall.cs#L487) causing the original client call to complete.
* Though it is not until a later in `GrpcCall.RunCall` where it calls `GrpcCall.FinishCall` which writes the stop event the OT instrumentation subscribes to [here](https://github.com/grpc/grpc-dotnet/blob/v2.25.x/src/Grpc.Net.Client/Internal/GrpcCall.cs#L662).

Starting and then stopping the test host immediately after the gRPC call seems to have resolved the flakiness. Though, I have a suspicion that all this is achieving is causing a long enough delay in the test for the stop event to occur in time. The test seems stable now, but perhaps there's a better way to ensure things happen in time.

Reply:
If you update to the latest version of Grpc.Net.Client then that is fixed. The diagnostic listener is written to BEFORE the result task is completed.

Reply:
v2.30 source:

https://github.com/grpc/grpc-dotnet/blob/v2.30.x/src/Grpc.Net.Client/Internal/GrpcCall.cs#L550-L562

Reply:
Cool. Yea it probably should have occurred to me to check out the current source...